src/kv/memstore.rs

```diff
@@ -1,55 +1,180 @@
+        return Ok(Box::new(ReadTransaction::new(self)));
+    }
+
+    async fn write<'a>(&'a self) -> Result<Box<dyn Write + 'a>> {
+        return Ok(Box::new(WriteTransaction::new(self)));
+    }
+
     async fn put(&mut self, key: &str, value: &[u8]) -> Result<()> {
-        self.map.insert(key.to_string(), value.to_vec());
+        self.map
+            .lock()
+            .await
+            .insert(key.to_string(), value.to_vec());
         Ok(())
     }
 
     async fn has(&self, key: &str) -> Result<bool> {
-        Ok(self.map.contains_key(key))
+        Ok(self.map.lock().await.contains_key(key))
+    }
+
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        match self.map.lock().await.get(key) {
+            None => Ok(None),
+            Some(v) => Ok(Some(v.to_vec())),
+        }
+    }
+}
+
+struct ReadTransaction<'a> {
+    store: &'a MemStore,
+}
+
+impl ReadTransaction<'_> {
+    fn new(store: &MemStore) -> ReadTransaction {
+        return ReadTransaction { store: store };
+    }
+}
+
+#[async_trait(?Send)]
+impl Read for ReadTransaction<'_> {
+    async fn has(&self, key: &str) -> Result<bool> {
+        Ok(self.store.map.lock().await.contains_key(key))
     }
 
     async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
-        match self.map.get(key) {
+        match self.store.map.lock().await.get(key) {
             None => Ok(None),
             Some(v) => Ok(Some(v.to_vec())),
         }
     }
 }
+
+struct WriteTransaction<'a> {
+    rt: ReadTransaction<'a>,
+    pending: Mutex<HashMap<String, Vec<u8>>>,
+    open: bool,
+}
+
+impl WriteTransaction<'_> {
+    fn new(store: &MemStore) -> WriteTransaction {
+        return WriteTransaction {
+            rt: ReadTransaction { store },
+            pending: Mutex::new(HashMap::new()),
+            open: true,
+        };
+    }
+
+    fn check_open(&self) -> Result<()> {
+        match self.open {
+            true => Ok(()),
+            false => Err(StoreError::Str("Transaction already closed".into())),
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Read for WriteTransaction<'_> {
+    async fn has(&self, key: &str) -> Result<bool> {
+        self.check_open()?;
+        match self.pending.lock().await.contains_key(key) {
+            true => Ok(true),
+            false => self.rt.has(key).await,
+        }
+    }
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        self.check_open()?;
+        match self.pending.lock().await.get(key) {
+            Some(v) => Ok(Some(v.to_vec())),
+            None => self.rt.get(key).await,
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Write for WriteTransaction<'_> {
+    async fn put(&self, key: &str, value: &[u8]) -> Result<()> {
+        self.check_open()?;
+        self.pending
+            .lock()
+            .await
+            .insert(key.to_string(), value.to_vec());
+        Ok(())
+    }
+
+    async fn commit(&mut self) -> Result<()> {
```

Comment:
Did you try having `commit` take `self` by value? In some iterations of my attempt, this was a nice way to prevent `commit` from getting called twice without needing `check_open`.

Reply:
No, but good instinct, and it's nice that it proves that the transaction isn't linked in anywhere either. Will try that out soon.